### PR TITLE
Data dictionary GA

### PR DESCRIPTION
**Pull request by zandre-eng**
## Technical Summary

Link to tickets [here](https://dimagi-dev.atlassian.net/browse/SC-2812) and [here](https://dimagi-dev.atlassian.net/browse/SC-2814).

This change moves the `DATA_DICTIONARY` feature flag to general availability for the Advanced Plan and higher. Furthermore, this PR also addressed hiding case property labels and descriptions for domains that do not have the `DATA_DICTIONARY` privilege.

This PR addresses one of several tickets that are aimed at moving the `DATA_DICTIONARY` feature flag to GA. This PR should therefore not be merged until blocking tickets have been finished, and the relevant public documentation has been created.

## Feature Flag

`DATA_DICTIONARY` which has been changed to a privilege.

### Safety story

- Local testing.
- QA testing done.
- Staging testing done.

### Automated test coverage

N/A

### QA Plan

QA testing done. Link to QA ticket [here](https://dimagi-dev.atlassian.net/browse/QA-5251).

### Migrations

- [ ] The migrations in this code can be safely applied first independently of the code

No, as the migration requires the created privilege to exist in the code to be able to bootstrap correctly.

### Rollback instructions

- [ ] This PR can be reverted after deploy with no further considerations

The following will need to be manually reverted when doing a rollback:
- The newly created Role (select * from django_prbac_role rle where rle.slug = 'data_dictionary' )
- The grant records created for this (select * from django_prbac_grant where to_role_id = {{ django_prbac_role id }})

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by SmittieC — COMMENTED** [review]
Overall looks fine to me. Is this being tested on staging?
Also, since this PR shouldn't be merged until all the tickets are done that's being blocked by this one, I recommend changing this PR to a draft. You don't have to, it's just to prevent accidental merge after approval :shrug: .

We'll probably need a temporary "master" branch for all the subsequent changes that needs this branch's changes (ze/data_dictionary_ga). This way we can just merge all progress into that one, and when all testing at the end of the day is done on staging, take that huge branch to master in one go. Toughts?

**Comment by zandre-eng** [comment]
@SmittieC This will be tested on staging, as well as taken through QA given the scope of this change. I do have the "awaiting QA" and "Open for review: Do not merge" labels set, so it shouldn't be possible to merge yet. I could still set it as draft however, just in case. Thanks for the suggestion :+1: 

That sounds like a good idea, I currently have two data dictionary PR's that have conflicts with each other. I think it'll make things a little easier merging everything to a single feature branch, fixing the conflicts there, and then push it together into master. Also that way, if there is any issues, all the changes can be easily reverted.

**Comment by Charl1996** [comment]
@zandre-eng Maybe merge master into this branch to get the linter updates. I see it's failing on rules which is supposed to be excluded.

**Comment by zandre-eng** [comment]
@Charl1996 Master has been merged in. :+1:

**Comment by kaapstorm** [comment]
Looks good to me.

I don't think this migration will be an issue. But we could split the first two commits out into a separate PR that can be merged and deployed first.

File: corehq/toggles/__init__.py
Comment (Charl1996) [inline corehq/toggles/__init__.py:2531]: Is this PR waiting on the public docs to be updated or will this be merged regardless?

Comment (zandre-eng) [reply corehq/toggles/__init__.py:2531]: Given the usage of this feature, I think we should wait until the relevant documentation has been updated. I'll then be able to update the help link in this PR and the documentation will be available to users once this change goes live.

**Comment by esoergel** [comment]
I haven't done a review of this PR, but I just wanted to flag that y'all might consider delaying the merge until July, as per Danny's recent email.  Very excited to see this coming, though!

File: corehq/apps/app_manager/app_schemas/tests/test_schema.py
Comment (kaapstorm) [inline corehq/apps/app_manager/app_schemas/tests/test_schema.py:26]: I haven't checked this, but I think from the traceback in the failing test, you need to patch 'corehq.apps.app_manager.app_schemas.casedb_schema.domain_has_privilege'

**Comment by kaapstorm** [comment]
I'm happy with those last three commits. It bothers me that the decorator isn't covered by tests (in fact none of those decorators are) but that is out of scope for this PR.

File: corehq/apps/app_manager/app_schemas/tests/test_schema.py
Comment (esoergel) [reply corehq/apps/app_manager/app_schemas/tests/test_schema.py:26]: FWIW there's also a `@privilege_enabled` decorator that may be useful here

Comment (zandre-eng) [reply corehq/apps/app_manager/app_schemas/tests/test_schema.py:26]: Thanks for the suggestions. @esoergel I unfortuantely wasn't able to get the `@privilege_enabled` decorator to work. I think that it is because the tests are subclassed from `SimpleTestCase`, and so are unable to run `domain_has_privilege()` correctly.

**Review by kaapstorm — APPROVED** [review]

**Comment by esoergel** [comment]
FYI I'm seeing a bug on the DD on prod and staging - the buttons to add new properties and groups don't appear to work.  See ticket here: https://dimagi-dev.atlassian.net/browse/SUPPORT-17029

**Comment by SmittieC** [comment]
We have to complete [this ticket](https://dimagi-dev.atlassian.net/browse/SC-2817) before merging this.

**Comment by zandre-eng** [comment]
@SmittieC I have updated the ticket to include the one you've linked as a blocker.

**Comment by esoergel** [comment]
Moving this to a draft PR as per this morning's discussion, indicating that we're waiting on a couple other things before this is considered ready to merge.

**Comment by esoergel** [comment]
@zandre-eng do y'all have a place for general feedback on the DD?  I keep noticing little things that would be nice to address as part of a GA effort, but I'm not sure where to provide feedback for triage.  For instance:

* There's no validation on case property names - you can use symbols, spaces, uppercase letters, whatever.  Though valid case property names in an application can't have any of that.
* There's similarly no validation on case type names
* Should `name` be added by default on call case types?
* There's no way to delete case types or properties (only deprecate them), even those that don't have any data associated with them
* You also can't edit or delete case properties that you just created and haven't yet saved - if you make a typo, the only option is to reload the page without saving and abandon your work in progress.

To be clear, I'm not saying these are things that need to be addressed, just wanted to be sure they were known.

**Comment by zandre-eng** [comment]
Hey @esoergel regarding where to give feedback for the DD, I think a comment on the [epic ticket](https://dimagi-dev.atlassian.net/browse/SC-2818) itself would probably be the best place. Thanks for raising this feedback, this is very helpful! I'll copy this onto the epic ticket and check in if these are things we'll address during the GA effort or afterwards.

**Review by SmittieC — APPROVED** [review]

**Pull request merged**
